tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/WindowAggExec.java

```diff
@@ -301,30 +301,327 @@ private void evaluationWindowFrame() {
     }
 
     for (int idx = 0; idx < functions.length; idx++) {
+      comp = null;
       if (orderedFuncFlags[idx]) {
-        comp = new BaseTupleComparator(inSchema, functions[idx].getSortSpecs());
-        Collections.sort(accumulatedInTuples, comp);
-        comp = new BaseTupleComparator(schemaForOrderBy, functions[idx].getSortSpecs());
+        SortSpec[] sortSpecs = functions[idx].getSortSpecs();
+        comp = new BaseTupleComparator(schemaForOrderBy, sortSpecs);
         Collections.sort(evaluatedTuples, comp);
+        // following comparator is used later when RANGE unit is handled to check whether order by value is changed or not
+        comp = new BaseTupleComparator(inSchema, sortSpecs);
+        Collections.sort(accumulatedInTuples, comp);
       }
 
-      for (int i = 0; i < accumulatedInTuples.size(); i++) {
-        Tuple inTuple = accumulatedInTuples.get(i);
-        Tuple outTuple = evaluatedTuples.get(i);
+      LogicalWindowSpec.LogicalWindowFrame.WindowFrameType windowFrameType = functions[idx].getLogicalWindowFrame().getFrameType();
+      WindowSpec.WindowFrameUnit windowFrameUnit = functions[idx].getLogicalWindowFrame().getFrameUnit();
+      int frameStart = 0, frameEnd = accumulatedInTuples.size() - 1;
+      int startOffset = functions[idx].getLogicalWindowFrame().getStartBound().getNumber();
+      int endOffset = functions[idx].getLogicalWindowFrame().getEndBound().getNumber();
+      functions[idx].bind(inSchema);
+
+      /*
+         Following code handles evaluation of window functions with two nested switch statements
+         Basically, ROWS handling has more cases then RANGE handling
+         First switch distinguishes among
+            1) built-in window functions without window frame support
+            2) buiit-in window functions with window frame support
+            3) aggregation functions, where window frame is supported
+         In window frame support case, there exists four types of window frame which is also handled by switch statement
+            a) Entire window partition
+            b) From the start of window partition to the moving end point relative to current row position
+            c) From the moving start point relative to current row position to the end of window partition
+            d) Both start point and end point of window frame are moving relative to the current row position
+
+         In the case of RANGE, there can be three window frame type
+            i) From the start of window partition to the last row that has the same order by key as the current row
+            ii) From the first row that has the same order by key as the current row to the end of window partition
+            iii) For all rows that has the same order by key as the current row
+       */
```

Comment:
Thanks for detailed and nice comments!
By the way, the below code block look quite complicated. Would you refactor it into several well-defined functions?